pkg/controller/nodelifecycle/node_lifecycle_controller_test.go

```diff
@@ -167,561 +166,685 @@ func newNodeLifecycleControllerFromClient(
-			t.Errorf("unexpected error: %v", err)
-		}
-		if err := nodeController.monitorNodeHealth(ctx); err != nil {
-			t.Errorf("unexpected error: %v", err)
-		}
-		zones := testutil.GetZones(item.fakeNodeHandler)
-		logger, _ := ktesting.NewTestContext(t)
-		for _, zone := range zones {
-			if _, ok := nodeController.zoneNoExecuteTainter[zone]; ok {
-				nodeController.zoneNoExecuteTainter[zone].Try(logger, func(value scheduler.TimedValue) (bool, time.Duration) {
-					nodeUID, _ := value.UID.(string)
-					pods, err := nodeController.getPodsAssignedToNode(value.Value)
-					if err != nil {
-						t.Errorf("unexpected error: %v", err)
-					}
-					t.Logf("listed pods %d for node %v", len(pods), value.Value)
-					controllerutil.DeletePods(ctx, item.fakeNodeHandler, pods, nodeController.recorder, value.Value, nodeUID, nodeController.daemonSetInformer.Lister())
-					return true, 0
-				})
-			} else {
-				t.Fatalf("Zone %v was uninitialized!", zone)
+	for testName, tt := range tests {
+		t.Run(testName, func(t *testing.T) {
+			_, ctx := ktesting.NewTestContext(t)
+			fakeNodeHandler := &testutil.FakeNodeHandler{
+				Existing:  tt.nodeList,
+				Clientset: fake.NewSimpleClientset(),
+			}
+			nodeController, _ := newNodeLifecycleControllerFromClient(
+				ctx,
+				fakeNodeHandler,
+				testRateLimiterQPS,
+				testRateLimiterQPS,
+				testLargeClusterThreshold,
+				testUnhealthyThreshold,
+				testNodeMonitorGracePeriod,
+				testNodeStartupGracePeriod,
+				testNodeMonitorPeriod)
+			nodeController.recorder = testutil.NewFakeRecorder()
+			nodeController.enterPartialDisruptionFunc = func(nodeNum int) float32 {
+				return testRateLimiterQPS
+			}
+			nodeController.enterFullDisruptionFunc = func(nodeNum int) float32 {
+				return testRateLimiterQPS
 			}
-		}
 
-		podEvicted := false
-		for _, action := range item.fakeNodeHandler.Actions() {
-			if action.GetVerb() == "delete" && action.GetResource().Resource == "pods" {
-				podEvicted = true
+			// initial zone state
+			nodeController.now = func() metav1.Time { return fakeNow }
+			if err := nodeController.syncNodeStore(fakeNodeHandler); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			if err := nodeController.monitorNodeHealth(ctx); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			if len(nodeController.zoneStates) != len(tt.expectedInitialStates) {
+				t.Errorf("unexpected length of initial zone state map: got %d instead of %d", len(nodeController.zoneStates), len(tt.expectedInitialStates))
+			}
+			for zone, state := range tt.expectedInitialStates {
+				if state != nodeController.zoneStates[zone] {
+					t.Errorf("unexpected initial zone state: %v: got %v instead of %v", zone, nodeController.zoneStates[zone], state)
+				}
 			}
-		}
 
-		if item.expectedEvictPods != podEvicted {
-			t.Errorf("expected pod eviction: %+v, got %+v for %+v", item.expectedEvictPods,
-				podEvicted, item.description)
-		}
+			// following zone state
```

Comment:
nit: do you want to use a for loop to de-dupe code of this section with previous "initial zone state" section?

Reply:
or a method.

Reply:
Just noticed that I missed this review from earlier commits - changed in a17b1aecbec